### PR TITLE
Range adaptors and begin end

### PR DESCRIPTION
**Pull request by ericniebler**
Implementation of [P0789](http://wg21.link/P0789), and the improved [range access] customization point objects.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,18 +13,18 @@ cache:
...
   include:
-    - env: GCC_VERSION=6 BUILD_TYPE=Debug ASAN=Off
-      os: linux
-      addons: &gcc6
-        apt:
-          packages:
-            - g++-6
-          sources:
-            - ubuntu-toolchain-r-test
+    # - env: GCC_VERSION=6 BUILD_TYPE=Debug ASAN=Off
```
Comment (CaseyCarter) [inline .travis.yml]: Please remove this commented-out GCC6 stuff.

Comment (ericniebler) [reply .travis.yml]: Will do. And don't merge yet. I'm bringing in the `subrange` changes from #126.

**Comment by ericniebler** [comment]
Will you want this squash merged? Or take all the commits? Or something else?

**Comment by CaseyCarter** [comment]
> Will you want this squash merged? Or take all the commits? Or something else?

Taking them all is fine.

**Pull request merged**